include/rmm/detail/logger.hpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// If using GCC, temporary workaround for older libcudacxx defining _LIBCPP_VERSION
+// undefine it before including spdlog, due to fmtlib checking if it is defined
+// TODO: remove once libcudacxx is on Github and RAPIDS depends on it
+#ifdef __GNUG__
+#undef _LIBCPP_VERSION
+#endif
+#include <spdlog/sinks/basic_file_sink.h>
+#include <spdlog/spdlog.h>
+
+#include <iostream>
+#include <string>
+
+namespace rmm {
+
+namespace detail {
+
+inline std::string default_log_filename()
+{
+  auto filename = std::getenv("RMM_DEBUG_LOG_FILE");
+  return (filename == nullptr) ? std::string{"rmm_log.txt"} : std::string{filename};
+}
+
+// Simple wrapper around a spdlog::logger that performs RMM-specific initialization
+struct logger_wrapper {
+  spdlog::logger logger_;
+
+  logger_wrapper()
+    : logger_{"RMM",
+              std::make_shared<spdlog::sinks::basic_file_sink_mt>(
+                default_log_filename(), true  // truncate file
+                )}
+  {
+    logger_.set_pattern("[%l][%t][%H:%M:%S:%f] %v");
```

Comment:
I don't understand why you can't sort by thread ID with the existing format?

Reply:
First, my suggestion is formatted incorrectly. The format I had in mind is:
```
[%t][%H:%M:%S:%f][%l] %v
```
It would be nice to open a log file in a text editor and "sort by lines ascending" to group the logs by thread. For that to work, the thread id would need to be printed first. The second part would need to be the timestamp, otherwise the logs would get out of order, even within a specific thread id. Therefore, the log level would need to be last. Since everything prior to the log level is fixed width, you also end up with a nice visual alignment.

```
[1232][23:59:58:056789][debug]
```

Reply:
In my experience grouping by thread has not been useful. You need to see what is happening in time order due to multiple threads using a resource. That said, I guess this change won't hurt anything.  I'll fix the visual alignment with an explicit width for the level.

Reply:
Done.